iml-ntp/main.rs

```diff
@@ -0,0 +1,147 @@
+use iml_wire_types::{ntp::{TimeSync, TimeStatus}, Fqdn};
+use std::io;
+
+/// This async function will retrieve the content type id of the specified model. A future
+/// containing the content type id of the model is returned.
+async fn get_managed_host_content_type_id(
+    mut client: Client,
+) -> Result<String, NtpError> {
+    client
+        .prepare("SELECT id FROM django_content_type WHERE model='managedhost';")
+        .compat()
+        .map_ok(move |statement| client.query(&statement, &[]).compat())
+        .try_flatten_stream()
+        .try_next()
+        .await?
+        .map(|row| row.get(0))
+        .ok_or_else(|| NtpError::IoError(io::Error::new(
+            io::ErrorKind::NotFound,
+            "managedhost",
+        )))
+}
+
+/// This async function will retrieve the managed host id for a given fqdn. A future
+/// containing the managed host id is returned.
+async fn get_managed_host_id(fqdn: &str, mut client: Client) -> Result<String, NtpError> {
+    client
+        .prepare("SELECT id FROM chroma_core_managedhost WHERE fqdn='$1';")
+        .compat()
+        .map_ok(move |statement| client.query(&statement, &[&fqdn]).compat())
+        .try_flatten_stream()
+        .try_next()
+        .await?
+        .map(|row| row.get(0))
+        .ok_or_else(|| NtpError::IoError(io::Error::new(io::ErrorKind::NotFound, fqdn)))
+}
+
+/// This async function will insert a new entry into the chroma_core_alertstate table. This will effectively raise an
+/// NtpOutOfSyncAlert alert for a given fqdn. 
+async fn add_alert(
+    fqdn: String,
+    alert_item_type_id: &str,
+    alert_item_id: &str,
+    mut client: Client,
+) -> Result<u64, NtpError> {
+    let start_time = Utc::now();
+    client
+        .prepare(
+            &format!(r#"INSERT INTO chroma_core_alertstate
+                (record_type, variant, alert_item_id, alert_type, begin, message, active, dismissed, severity, alert_item_type_id) values
+                ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10);"#
+            )
+        )
+        .compat()
+        .and_then(move |s| {
+            client.execute(&s, &[&"NtpOutOfSyncAlert", &"{}", &alert_item_id, &"NtpOutOfSyncAlert", &&format!("{}", start_time).as_str(), &format!("Ntp is not in sync on server {}", fqdn).as_str(), &true, &false, &40, &alert_item_type_id]).compat()
+        })
+        .map(|x| Ok(x))
+        .map_err(|e: tokio_postgres::error::Error| NtpError::IoError(io::Error::new(io::ErrorKind::Other, e.to_string())))
+        .await?
+        .map_err(|e: tokio_postgres::error::Error| NtpError::IoError(io::Error::new(io::ErrorKind::Other, e.to_string())))
+}
+
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn std::error::Error>> {
+    env_logger::init();
+
+    loop { 
+        log::info!("Entering loop and connecting to rabbit.");
+        let client = connect_to_rabbit().compat().await?;
+
+        tokio::spawn(async move {
```

Comment:
likewise I don't think you should be spawning to a new thread here. 

You'll basically want to pull from the service queue in the main thread.

Reply:
ok, maybe that's why i'm getting an access error. I'll update.